scrapy/utils/datatypes.py

```diff
@@ -69,16 +70,35 @@ class LocalCache(collections.OrderedDict):
     Older items expires first.
     """
 
-    def __init__(self, limit=None):
+    def __init__(self, limit=None, time_limit=None):
         super().__init__()
         self.limit = limit
+        self.time_limit = time_limit
 
     def __setitem__(self, key, value):
-        if self.limit:
+        now = datetime.now()
+        value = (value, now)
+
+        if not self.limit is None:
             while len(self) >= self.limit:
                 self.popitem(last=False)
+            for _, val in self.items():
+                if not time_limit is None and int((now - val[1]).total_seconds()) > self.time_limit:
+                    self.popitem(last=False)
+                else:
+                    break
         super().__setitem__(key, value)
 
+    def __getitem__(self, key):
+        for _, val in self.items():
+            now = datetime.now()
+            if not time_limit is None and int((now - val[1]).total_seconds()) > self.time_limit:
```

Comment:
for line 95, `if not time_limit is None`, here do you mean `self.time_limit` ? If yes, then you should change this line to `if self.time_limit is not None` and move this part upper before for loop to decrease compare times to increase.

```python
def __getitem__(self, key):
        for _, val in self.items():
            if self.time_limit is None:
                return super().__getitem__(key)[0]
            now = datetime.now()
            if (now - val[1]).total_seconds() > self.time_limit:
                self.popitem(last=False)
```

Reply:
I'm dazzled that this worked locally before, I must have not tested it properly before; Also, good performance catch.
I applied your suggestion.